### PR TITLE
OAK-8898: On querying, IndexReader failed with AlreadyClosedException

### PR DESCRIPTION
**Pull request by tihom88**

### CODE REVIEW

File: oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/update/TimedRefreshPolicy.java [hunk 1]
```diff
@@ -53,16 +54,20 @@ public void refreshOnWriteIfRequired(Runnable refreshCallback) {
...
-            long currentTime = clock.getTime();
-            if (currentTime - lastRefreshTime > refreshDelta
-                    && dirty.compareAndSet(true, false)){
-                lastRefreshTime = currentTime;
-                refreshCallback.run();
+        synchronized (lock) {
+            if (dirty.get()) {
+                long currentTime = clock.getTime();
+                if (currentTime - lastRefreshTime > refreshDelta
+                        && dirty.compareAndSet(true, false)) {
```
Comment (averma21) [inline oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/update/TimedRefreshPolicy.java]: Could you please explain what is the intention here. Do we want that dirty stays `false` while `refreshCallback` is running?

File: oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/update/TimedRefreshPolicy.java [hunk 2]
```diff
@@ -53,16 +54,20 @@ public void refreshOnWriteIfRequired(Runnable refreshCallback) {
...
     }
 
     private void refreshIfRequired(Runnable refreshCallback) {
-        if (dirty.get()){
-            long currentTime = clock.getTime();
-            if (currentTime - lastRefreshTime > refreshDelta
-                    && dirty.compareAndSet(true, false)){
-                lastRefreshTime = currentTime;
-                refreshCallback.run();
+        synchronized (lock) {
```
Comment (averma21) [inline oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/update/TimedRefreshPolicy.java]: Could you please explain what is the intention behind synchronizing on `lock` . Do we want that dirty stays `false` while `refreshCallback` is running?

Comment (thomasmueller) [reply oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/update/TimedRefreshPolicy.java]: I think it's not needed to synchronize updated(). It should be enough to synchronize just the call to refreshCallback.run() in my view.

File: oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/LuceneIndexNodeManager.java [hunk 3]
```diff
@@ -272,7 +271,9 @@ private void closeReaders(Iterable<LuceneIndexReader> readers) {
     }
 
     private void releaseHolder(SearcherHolder holder) {
-        decrementSearcherUsageCount(holder.searcher);
+        synchronized (holder) {
```
Comment (thomasmueller) [inline oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/LuceneIndexNodeManager.java]: Shouldn't we also synchronize closeReaders(), together with releaseHolder(..)?

File: oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/update/TimedRefreshPolicy.java [hunk 4]
```diff
@@ -21,6 +21,7 @@
 
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
```
Comment (averma21) [inline oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/update/TimedRefreshPolicy.java]: This `AtomicLong` import is not needed now.

**Comment by tihom88** [comment]
Looks like we were looking at this in a wrong way. What we want is to change it in such a way so that release holder in not called twice on a holder. So removed all locks and instead of blocking thread now using an atomic boolean to make threads bypass decrementing references of index reader.

**Pull request closed without merging**